acvm-repo/brillig/src/opcodes.rs

```diff
@@ -84,6 +84,25 @@ impl HeapValueType {
     pub fn field() -> HeapValueType {
         HeapValueType::Simple(BitSize::Field)
     }
+
+    /// Returns the total number of field elements required to represent this type in memory
+    pub fn flattened_size(&self) -> usize {
+        match self {
+            HeapValueType::Simple(_) => 1,
+
+            HeapValueType::Array { value_types, size } => {
+                // Recursively compute the flattened size of all value types
+                let element_size: usize = value_types.iter().map(|ty| ty.flattened_size()).sum();
+
+                element_size * size
+            }
+
+            // The size of a vector is stored separately and should be handled by the caller of this method
+            HeapValueType::Vector { value_types } => {
+                value_types.iter().map(|ty| ty.flattened_size()).sum()
+            }
+        }
+    }
```

Comment:
This is a bit wonky as it's returning the full type size for simple and array types but only the element size for vectors. I've had to implement a similar method in a different PR but I return `None` here if we encounter a vector.

Reply:
I'll use `Either` for it to make a distinction.

Reply:
Or maybe just move this function as a private one into the module where it's called, as an Array of Vector is not handled by it.